c10/util/order_preserving_flat_hash_map.h

```diff
@@ -843,6 +878,76 @@ class sherwood_v3_table : private EntryAlloc, private Hasher, private Equal
         swap(num_elements, other.num_elements);
         swap(max_lookups, other.max_lookups);
         swap(_max_load_factor, other._max_load_factor);
+        swap(sentinel, other.sentinel);
+        swap(sentinel_val, other.sentinel_val);
+    }
+
+    void reset_list() {
+      sentinel->next = sentinel;
+      sentinel->prev = sentinel;
+    }
+
+    void remove_from_list(EntryPointer elem) {
+      elem->prev->next = elem->next;
+      elem->next->prev = elem->prev;
+    }
+
+    void insert_after(EntryPointer new_elem, EntryPointer prev) {
+      auto next = prev->next;
+
+      prev->next = new_elem;
+      new_elem->prev = prev;
+
+      new_elem->next = next;
+      next->prev = new_elem;
+    }
+
+    void swap_adjacent_nodes(EntryPointer before, EntryPointer after) {
+      // sentinel stays consant, so before->prev cannot equal after
+      auto before_prev = before->prev;
+      auto after_next = after->next;
+
+      before_prev->next = after;
+      after->prev = before_prev;
+
+      after_next->prev = before;
+      before->next = after_next;
+
+      before->prev = after;
+      after->next = before;
+    }
+
+    void swap_positions(EntryPointer p1, EntryPointer p2) {
```

Comment:
What is going on here? before this operation was just composed of two already-existing functions. Why did it get so much more complicated?